// javascript-serializer/dist/src/convertors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var errio_1 = require("errio");
var SerializableError = /** @class */ (function () {
    function SerializableError(error, options) {
        if (options === void 0) { options = {}; }
        var _this = this;
        this.error = error;
        this.options = options;
        this.toJSON = function () { return errio_1.toObject(_this.error, { stack: _this.options.stack !== false }); };
    }
    SerializableError.fromJSON = function (obj) { return errio_1.fromObject(obj); };
    return SerializableError;
}());
var SerializableRegExp = /** @class */ (function () {
    function SerializableRegExp(regexp, options) {
        this.regexp = regexp;
        this.options = options;
    }
    SerializableRegExp.prototype.toJSON = function () {
        var _a = this.regexp, source = _a.source, flags = _a.flags, lastIndex = _a.lastIndex;
        return { source: source, flags: flags, lastIndex: lastIndex };
    };
    SerializableRegExp.fromJSON = function (_a) {
        var source = _a.source, flags = _a.flags, lastIndex = _a.lastIndex;
        var regexp = new RegExp(source, flags);
        regexp.lastIndex = lastIndex;
        return regexp;
    };
    return SerializableRegExp;
}());
exports.convertors = [
    { from: Error, to: SerializableError },
    { from: RegExp, to: SerializableRegExp },
];
